### PR TITLE
DM-32894: add support for data clients

### PR DESCRIPTION
**Pull request by r-owen**

### CODE REVIEW

**Review by wvreeven — APPROVED** [review]
LGTM

**Comment by tribeiro** [comment]
I would gently suggest that you try to maintain backward compatibility with the current production cycle.. Breaking current cycle compatibility means we won't be able to easily deploy this at the summit anytime soon.

**Review by wvreeven — CHANGES_REQUESTED** [review]
One unit test fails. Please check.

**Comment by r-owen** [comment]
@tribeiro this change is backwards compatible with ts_xml but not ts_idl. I have no idea how to make it backwards compatible with ts_idl, since the error codes have almost no overlap.

**Comment by tribeiro** [comment]
I see.. the problem is that idl is part of the cycle definition so, upgrading it won't allow us to easily deploy it. 

If I may suggest something, you could add the error code to the current code and mark it for deprecation once the new idl is released. I know it is ugly but it would allow us to retain current cycle compatibility.

File: python/lsst/ts/ess/csc/ess_csc.py [hunk 1]
```diff
@@ -22,14 +22,16 @@
...
+from typing import Dict, List, Optional, Sequence, Tuple, Union
+
+import jsonschema
 
 from .config_schema import CONFIG_SCHEMA
 from . import __version__
-from lsst.ts import salobj, tcpip, utils
+from lsst.ts import salobj, utils
 from lsst.ts.ess import common
 from lsst.ts.idl.enums.ESS import ErrorCode
```
Comment (tribeiro) [inline python/lsst/ts/ess/csc/ess_csc.py:36]: Here is a simple alternative to make the code backwards compatible with the current cycle:

```
try:
    from lsst.ts.idl.enums.ESS import ErrorCode
except ImportError:
    warning.warn("Running in ts-idl < 3.5 compatibility mode....")
    class ErrorCode(IntEnum):
        """Error codes that can be used by the ESS CSC."""


        NotConnected = 1
        AlreadyConnected = 2
        NotConfigured = 3
        BadConfiguration = 4
        ReadLoopFailed = 5
        TelemetryError = 6
        HardwareError = 7
```

Comment (tribeiro) [reply python/lsst/ts/ess/csc/ess_csc.py:36]: Of course, this is not going to work because what changed was the enum itself...

Comment (tribeiro) [reply python/lsst/ts/ess/csc/ess_csc.py:36]: But could be something on these lines..  You could even do something like:

```
# TODO: Use enum from ts-idl when version >3.7 is released.
class ErrorCode(IntEnum):
    """Error codes that can be used by the ESS CSC.

    Values:

    * ConnectionFailed: a model could not connect its data server.
      Perhaps the model configuration is wrong or the server is down.
    * ConnectionLost: a model lost its connection to its data server.
    * StartFailed: one or more models failed to start.
    * RunFailed: one or more models failed while running.
    """
    ConnectionFailed = 1
    ConnectionLost = 2
    StartFailed = 3
    RunFailed = 4
```

Anything to maintain backward compatibility with a newly released cycle would be welcomed!

Comment (r-owen) [reply python/lsst/ts/ess/csc/ess_csc.py]: @tribeiro yes I did as you suggested (before seeing your suggestion): I added a local ErrorCode class to ts_ess_csc and filed a ticket to remove it when ts_idl 3.7 is available.

**Comment by r-owen** [comment]
For the record: the test failure was triggered when mypy checked doc/conf.py (though that was not at all clear from the output) -- in particular when that file imported lsst.ts.ess.csc. It appears to be something about the implicit lsst.ts.ess namespace. In any case the fix I settled on is to put that import in a `if not typing.TYPE_CHECKING:` block.

**Comment by r-owen** [comment]
I overhauled the docs and fixed old references to ts_ess in cond/meta.yaml, Jenksinsfile.conda, and README.rst. If you have time I'd appreciate it if you could have another look -- focusing on those changes. I also just realized I forgot to add the dynamic loading, but I can do that as part of the labjack update.

File: Jenkinsfile.conda [hunk 2]
```diff
@@ -1,2 +1,2 @@
 @Library('JenkinsShared')_
-CondaPipeline(["ts_config_ocs"], "ts-ess", "lsst.ts.ess", "noarch")
+CondaPipeline(["ts_config_ocs"], "ts-ess-csc", "lsst.ts.ess.csc", "noarch")
```
Comment (wvreeven) [inline Jenkinsfile.conda:2]: Please revert this change. The cycle build expects the conda package to be called `ts-ess`. I created DM-32978 for doing this since it requires more changes than just this file.

File: README.rst [hunk 3]
```diff
@@ -4,7 +4,7 @@ ts_ess_csc
 
 ``ts_ess_csc`` is a Commandable SAL Component (CSC) to control various environmental sensors at the Vera C. Rubin Observatory.
 
-`Documentation <https://ts-ess.lsst.io>`_
+`Documentation <https://ts-ess-csc.lsst.io>`_
```
Comment (wvreeven) [inline README.rst:7]: Thanks for catching that!

File: conda/meta.yaml [hunk 4]
```diff
@@ -1,7 +1,7 @@
 {% set data= load_setup_py_data() %}
 
 package:
-  name: "ts-ess"
+  name: "ts-ess-csc"
```
Comment (wvreeven) [inline conda/meta.yaml:4]: Please revert this chnage as well as explained before.

File: doc/index.rst [hunk 5]
```diff
@@ -6,42 +6,95 @@
...
+    :target: https://jira.lsstcorp.org/issues/?jql=labels+%3D+ts_ess_csc
 
-.. _ts_ess_controller: https://ts-ess_controller.lsst.io
-.. _ts_ess_common: https://ts-ess_common.lsst.io
+Overview
+========
 
-.. _lsst.ts.ess-using:
+The ESS Commandable SAL Component (CSC) reads various environmental sensors at the Vera C. Rubin Observatory,
+and publishes the resulting data in ESS telemetry topics.
```
Comment (wvreeven) [inline doc/index.rst:22]: Please use one line per sentence and one sentence per line.

File: doc/index.rst [hunk 6]
```diff
@@ -6,42 +6,95 @@
...
+-------------
+
+Configuration files are stored in `ts_config_ocs`_.
+
+There should be one standard configuration file for each site.
+Each configuration file specifies configuration for all ESS SAL indices supported at that site.
+To run all ESS CSCs appropriate for a site, examine the configuration file and run one ESS CSC for each entry in it.
+
+The ESS CSC uses "data clients" to communicate with environmental data servers and publish the telemetry,
+in order to flexibly handle different kinds of servers.
```
Comment (wvreeven) [inline doc/index.rst:48]: Please use one line per sentence and one sentence per line.

File: doc/version_history.rst [hunk 7]
```diff
@@ -1,11 +1,32 @@
...
+* Use new error codes from ts_idl 3.7, which is recommended but not required, due to a temporary local version of the ErrorCode enum class.
+  All clients of this CSC should use ts_idl v3.7.0 in order to get correct ErrorCode values.
+* Fix API docs.
+* Enable mypy type checking.
+* Change ``master`` to ``main`` in CONFIG_SCHEMA's ``id``, in preparation for renaming the branch.
+
+Requires:
+
+* ts_salobj 6.3
+* ts_idl 3.7
```
Comment (wvreeven) [inline doc/version_history.rst:23]: This version of ts_idl is not required, right?

**Review by wvreeven — CHANGES_REQUESTED** [review]
Please revert the changes to the conda build files. As commented, this requires more changes than in this project alone and I created a ticket for doing that.

File: Jenkinsfile.conda (hunk 2 above)
Comment (r-owen) [reply Jenkinsfile.conda]: Done. Thanks for filing a ticket for this!

File: conda/meta.yaml (hunk 4 above)
Comment (r-owen) [reply conda/meta.yaml]: Done

File: doc/index.rst (hunk 6 above)
Comment (r-owen) [reply doc/index.rst:48]: OK.

File: doc/index.rst (hunk 5 above)
Comment (r-owen) [reply doc/index.rst:22]: OK

File: doc/version_history.rst (hunk 7 above)
Comment (r-owen) [reply doc/version_history.rst]: Good point. I modified the text to recommend 3.7 but allow 3.5 and 3.6

**Review by wvreeven — APPROVED** [review]
Looks great now. Thanks very much.
